### PR TITLE
Tracking doc: rewrite `@wdio/local-runner` package into TypeScript

### PR DESCRIPTION
**Pull request by christian-bromann**
This is a tracking issue for organising the effort of re-writing the code base from vanilla JS into TypeScript. If you are reading this and are interested getting involved, this is a great chance to do so. Here are the rules:

- if you are interested converting source files from `/packages/wdio-local-runner` into TypeScript please comment on this thread and mention the files you want to convert to avoid duplicate work
- there is no reason to convert the whole package at once - smaller PRs make it easier to review the code
- if you make a PR, reference this issue so we can keep track of the progress
- if you make a PR, label it with `TypeScript` and the package name `wdio-local-runner` so we can easier find all PRs
- if you make a PR, assign it to the `TypeScript Rewrite` project

Happy rewriting 😅

### CODE REVIEW

File: packages/wdio-local-runner/src/index.ts [hunk 1]
```diff
@@ -3,14 +3,25 @@ import { WritableStreamBuffer } from 'stream-buffers'
...
+    args: any
+}
+
 export default class LocalRunner {
-    constructor (configFile, config) {
-        this.configFile = configFile
+    config: WebdriverIO.Config
+    workerPool: Record<string, WorkerInstance> = {}
+
+    stdout: WritableStreamBuffer
```
Comment (erwinheitzman) [inline packages/wdio-local-runner/src/index.ts:19]: Also check if you can remove the type 👍 
```suggestion
    stdout: WritableStreamBuffer = new WritableStreamBuffer(BUFFER_OPTIONS)
```

File: packages/wdio-local-runner/src/index.ts [hunk 2]
```diff
@@ -3,14 +3,25 @@ import { WritableStreamBuffer } from 'stream-buffers'
...
+    stderr: WritableStreamBuffer
```
Comment (erwinheitzman) [inline packages/wdio-local-runner/src/index.ts:20]:
```suggestion
    stderr: WritableStreamBuffer = new WritableStreamBuffer(BUFFER_OPTIONS)
```

File: packages/wdio-local-runner/src/index.ts [hunk 3]
```diff
@@ -3,14 +3,25 @@ import { WritableStreamBuffer } from 'stream-buffers'
...
+
+    constructor (configFile: string, config: WebdriverIO.Config) {
         this.config = config
-        this.workerPool = {}
+
         this.stdout = new WritableStreamBuffer(BUFFER_OPTIONS)
```
Comment (erwinheitzman) [inline packages/wdio-local-runner/src/index.ts:25]: Remove these after making the suggested changes

File: packages/wdio-local-runner/src/replQueue.ts [hunk 4]
```diff
@@ -1,16 +1,27 @@
...
+    onStart: Function
+    onEnd: Function
+}
+
 /**
  * repl queue class
  * allows to run debug commands in mutliple workers one after another
  */
 export default class ReplQueue {
+    repls: Repl[]
```
Comment (erwinheitzman) [inline packages/wdio-local-runner/src/replQueue.ts:17]:
```suggestion
    repls: Repl[] = []
```

File: packages/wdio-local-runner/src/replQueue.ts [hunk 5]
```diff
@@ -1,16 +1,27 @@
...
+    runningRepl?: WDIORepl
+
     constructor () {
-        this.runningRepl = null
         this.repls = []
```
Comment (erwinheitzman) [inline packages/wdio-local-runner/src/replQueue.ts:21]: remove this after making the suggested change

File: packages/wdio-local-runner/src/replQueue.ts [hunk 6]
```diff
@@ -19,16 +30,16 @@ export default class ReplQueue {
             return
         }
 
-        const { childProcess, options, onStart, onEnd } = this.repls.shift()
-        this.runningRepl = new WDIORepl(childProcess, options)
+        const { childProcess, options, onStart, onEnd } = this.repls.shift() || ({} as Repl)
```
Comment (erwinheitzman) [inline packages/wdio-local-runner/src/replQueue.ts:33]: The or, should not be needed I think?
```suggestion
        const { childProcess, options, onStart, onEnd } = this.repls.shift()
```

File: packages/wdio-local-runner/src/run.ts [hunk 7]
```diff
@@ -7,30 +7,38 @@ import { SHUTDOWN_TIMEOUT } from './constants'
 
 const log = logger('@wdio/local-runner')
 
-export const runner = new Runner()
+/**
+ * ToDo(Christian): remove when @wdio/runner got typed
+ */
+interface RunnerInterface extends NodeJS.EventEmitter {
+    sigintWasCalled: boolean
+    [key: string]: any
```
Comment (erwinheitzman) [inline packages/wdio-local-runner/src/run.ts:15]: This line can be removed as it already extends the object, so there's no need to tell TS that there will be more properties

Comment (christian-bromann) [reply packages/wdio-local-runner/src/run.ts:15]: a runner instance (which is a child of eventemitter) can have arbitrary commands which is why we actually need this, particularly in the line where we call that arbitrary command:

```ts
runner[m.command as string](m).then(
```

File: packages/wdio-local-runner/src/worker.ts [hunk 8]
```diff
@@ -108,12 +134,12 @@ export default class WorkerInstance extends EventEmitter {
         /**
          * handle debug command called within worker process
          */
-        if (payload.origin === 'debugger' && payload.name === 'start') {
+        if (childProcess && payload.origin === 'debugger' && payload.name === 'start') {
```
Comment (erwinheitzman) [inline packages/wdio-local-runner/src/worker.ts:137]:
```suggestion
        if (payload?.origin === 'debugger' && payload?.name === 'start') {
```

File: packages/wdio-local-runner/tests/__mocks__/@wdio/local-runner.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
+import WorkerInstance from '../../../src/worker'
+
+export default class LocalRunnerMock {
+    configFile: string
+    config: WebdriverIO.Config
+    workerPool: Record<string, WorkerInstance>
```
Comment (erwinheitzman) [inline packages/wdio-local-runner/tests/__mocks__/@wdio/local-runner.ts:6]:
```suggestion
    workerPool: Record<string, WorkerInstance> = {}
```

File: packages/wdio-local-runner/tests/__mocks__/@wdio/local-runner.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    initialise: Function
```
Comment (erwinheitzman) [inline packages/wdio-local-runner/tests/__mocks__/@wdio/local-runner.ts:7]:
```suggestion
    initialise: Function = jest.fn()
```

File: packages/wdio-local-runner/tests/__mocks__/@wdio/local-runner.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+    constructor (configFile: string, config: WebdriverIO.Config) {
+        this.configFile = configFile
+        this.config = config
+        this.workerPool = {}
```
Comment (erwinheitzman) [inline packages/wdio-local-runner/tests/__mocks__/@wdio/local-runner.ts:12]:
```suggestion


```

File: packages/wdio-local-runner/tests/__mocks__/@wdio/local-runner.ts (hunk 10 above)
Comment (erwinheitzman) [reply packages/wdio-local-runner/tests/__mocks__/@wdio/local-runner.ts:7]: Remove in constructor

File: packages/wdio-local-runner/tests/__mocks__/@wdio/local-runner.ts (hunk 9 above)
Comment (erwinheitzman) [reply packages/wdio-local-runner/tests/__mocks__/@wdio/local-runner.ts:6]: Remove in constructor

**Review by erwinheitzman — CHANGES_REQUESTED** [review]
Some small change requests

File: packages/wdio-local-runner/src/run.ts [hunk 12]
```diff
@@ -7,30 +7,38 @@ import { SHUTDOWN_TIMEOUT } from './constants'
...
 runner.on('exit', process.exit.bind(process))
-runner.on('error', ({ name, message, stack }) => process.send({
+runner.on('error', ({ name, message, stack }) => process.send!({
     origin: 'worker',
     name: 'error',
     content: { name, message, stack }
 }))
 
-process.on('message', (m) => {
+process.on('message', (m: any) => {
```
Comment (mgrybyk) [inline packages/wdio-local-runner/src/run.ts:26]: Can you add types for a message?

**Review by mgrybyk — CHANGES_REQUESTED** [review]
One minor comment, LGTM overall 

there are still some comments from @erwinheitzman

File: packages/wdio-local-runner/src/worker.ts (hunk 8 above)
Comment (christian-bromann) [reply packages/wdio-local-runner/src/worker.ts:137]: it's

`childProcess && payload.origin === 'debugger'`
not
`payload && payload.origin === 'debugger'`

**Comment by christian-bromann** [comment]
@mgrybyk @erwinheitzman addressed all comments, good to go?

File: packages/wdio-local-runner/src/index.ts [hunk 13]
```diff
@@ -3,16 +3,24 @@ import { WritableStreamBuffer } from 'stream-buffers'
...
 export default class LocalRunner {
-    constructor (configFile, config) {
-        this.configFile = configFile
+    config: WebdriverIO.Config
+    workerPool: Record<string, WorkerInstance> = {}
+
+    stdout = new WritableStreamBuffer(BUFFER_OPTIONS)
+    stderr = new WritableStreamBuffer(BUFFER_OPTIONS)
+
+    constructor (configFile: string, config: WebdriverIO.Config) {
```
Comment (erwinheitzman) [inline packages/wdio-local-runner/src/index.ts:22]: Just noticed that previously we assigned the configFile which is now missing, is this an oopsy?

Comment (erwinheitzman) [reply packages/wdio-local-runner/src/index.ts:22]: Last time I forgot you can also do:
```suggestion
    constructor (public configFile: string, public config: WebdriverIO.Config) {
```
Which allows you to remove the declarations above for these two

Comment (christian-bromann) [reply packages/wdio-local-runner/src/index.ts:22]: > is this an oopsy?

no, it was there before but not used in the local runner (I think it was in the lambda one). Happy to clean this up eventually as part of a separate PR.

**Review by erwinheitzman — APPROVED** [review]
LGTM

**Pull request merged**